src/adapter/src/catalog.rs

```diff
@@ -5360,6 +5379,103 @@ impl Catalog {
                     }
                 }
+                Op::ClusterOwner { id, new_owner } => {
+                    let cluster_name = state.get_cluster(id).name().to_string();
+                    if id.is_system() {
+                        return Err(AdapterError::Catalog(Error::new(
+                            ErrorKind::ReadOnlyCluster(cluster_name),
+                        )));
+                    }
+                    builtin_table_updates.push(state.pack_cluster_update(&cluster_name, -1));
+                    let cluster = state.get_cluster_mut(id);
+                    cluster.owner_id = new_owner;
+                    tx.update_cluster(id, cluster)?;
+                    builtin_table_updates.push(state.pack_cluster_update(&cluster_name, 1));
+                }
+                Op::ClusterReplicaOwner {
+                    cluster_id,
+                    replica_id,
+                    new_owner,
+                } => {
+                    let cluster = state.get_cluster(cluster_id);
+                    if cluster_id.is_system() {
+                        return Err(AdapterError::Catalog(Error::new(
+                            ErrorKind::ReadOnlyCluster(cluster.name().to_string()),
+                        )));
+                    }
+                    let replica_name = cluster
+                        .replicas_by_id
+                        .get(&replica_id)
+                        .expect("catalog out of sync")
+                        .name
+                        .clone();
+                    builtin_table_updates.push(state.pack_cluster_replica_update(
+                        cluster_id,
+                        &replica_name,
+                        -1,
+                    ));
+                    let cluster = state.get_cluster_mut(cluster_id);
+                    let replica = cluster
+                        .replicas_by_id
+                        .get_mut(&replica_id)
+                        .expect("catalog out of sync");
+                    replica.owner_id = new_owner;
+                    tx.update_cluster_replica(cluster_id, replica_id, replica)?;
+                    builtin_table_updates.push(state.pack_cluster_replica_update(
+                        cluster_id,
+                        &replica_name,
+                        1,
+                    ));
+                }
+                Op::DatabaseOwner { id, new_owner } => {
+                    let database = state.get_database(&id);
+                    builtin_table_updates.push(state.pack_database_update(database, -1));
+                    let database = state.get_database_mut(&id);
+                    database.owner_id = new_owner;
+                    let database = state.get_database(&id);
+                    tx.update_database(id, database)?;
+                    builtin_table_updates.push(state.pack_database_update(database, 1));
+                }
+                Op::SchemaOwner {
```

Comment:
Does this need to protect against system schemas?

Reply:
Though if you think I should make some specific changes regarding the validations to this PR I'm happy to do it.